### PR TITLE
[String] Grapheme fast paths for punctuation: 5-8x speedup.

### PR DESCRIPTION
**Pull request by milseman**
Many strings use non-sub-300 punctuation characters (e.g. unicode
hyphen, CJK quotes, etc). This can cause switching between fast and
slow paths for grapheme breaking. Add in fast-paths for general
punctuation characters and CJK punctuation and symbol characters.

This results in about a 5-8x speedup for heavily (unicode) punctuated
Latiny and CJKy workloads.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@shahmishal Bots are having issues:
19:41:38 Caused: java.io.IOException: Backing channel 'macOS-30' is disconnected.

**Comment by milseman** [comment]
@swift-ci please smoke test

**Comment by milseman** [comment]
@swift-ci please benchmark

**Review by xwu — COMMENTED** [review]
Nit: the TODO comment [here](https://github.com/apple/swift/pull/10648/files#diff-e3e73041cb4200c0c609eaf60d5c2754R327) and the one above about CJK punctuation can go now, yes?

**Comment by milseman** [comment]
Yup

**Comment by shahmishal** [comment]
> @shahmishal Bots are having issues:
> 19:41:38 Caused: java.io.IOException: Backing channel 'macOS-30' is disconnected.

We had network outage.

**Comment by milseman** [comment]
@swift-ci please test

File: stdlib/public/core/StringCharacterView.swift [hunk 1]
```diff
@@ -344,6 +341,14 @@ extension String.CharacterView : BidirectionalCollection {
       //   0xAC00–0xD7AF
       case 0xac00...0xd7af: return true
 
+      // Common general use punctuation, excluding extenders:
+      //   0x2010-0x2029
+      case 0x2010...0x2029: return true
+
+      // CJK punctuation characters, excluding extenders:
```
Comment (jrose-apple) [inline stdlib/public/core/StringCharacterView.swift]: What do you mean by "extenders"?

Comment (milseman) [inline stdlib/public/core/StringCharacterView.swift]: The characters (i.e. scalars) that have the "extend" property for the purposes of grapheme breaking. Such characters usually* don't have a grapheme break before them. E.g. 0x302A.

See:
http://unicode.org/reports/tr29/#GB9
http://www.unicode.org/Public/UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt

**Pull request merged**